### PR TITLE
Parse signaling messages in P2PClient.

### PR DESCRIPTION
**Pull request by jianjunz**
P2PClient scans signaling messages to see if a PeerConnectionChannel is needed to be created or not. However, scaning messages with std::string::find may miss some matches when extra spaces are added, although these spaces don't change the meaning of a JSON message, e.g.: `{"type":"chat-closed"}` is the same as `{ "type" : "chat-closed" }`. Parsing signaling messages in P2PClient identifies message type more accurate.

### CODE REVIEW

File: talk/owt/sdk/p2p/p2pclient.cc [hunk 1]
```diff
@@ -251,79 +256,88 @@ void P2PClient::OnSignalingMessage(const std::string& message,
...
+          // If our ID is larger than remote, make the remote side as the
+          // publisher (offerer) In case we already have an offer. So we
+          // remove current PCC and create answer.
           std::shared_ptr<LocalStream> stream = pcc->GetLatestLocalStream();
           std::function<void()> success_callback =
               pcc->GetLatestPublishSuccessCallback();
           std::function<void(std::unique_ptr<Exception>)> failure_callback =
               pcc->GetLatestPublishFailureCallback();
-          pcc->SetAbandoned();
+          pcc->Stop(nullptr, nullptr);
```
Comment (jianjunz) [inline talk/owt/sdk/p2p/p2pclient.cc:289]: The logical for "offer" is unchanged. Old version (line 272) uses pcc->Stop as well. Not sure why it's identified as a change.

File: talk/owt/sdk/p2p/p2pclient.cc [hunk 2]
```diff
@@ -251,79 +256,88 @@ void P2PClient::OnSignalingMessage(const std::string& message,
...
-        Json::Value stop_info;
-        rtc::GetValueFromJsonObject(json_message, "data", &stop_info);
-        rtc::GetIntFromJsonObject(stop_info, "code", &code);
-        rtc::GetStringFromJsonObject(stop_info, "message", &error);
-
-        if (code == kWebrtcIceGatheringPolicyUnsupported) {
-          auto pcc = that->GetPeerConnectionChannel(remote_id);
+    } else if (message_type == "chat-signal"){
+      Json::Value signal;
+      rtc::GetValueFromJsonObject(json_message, kMessageDataKey, &signal);
```
Comment (taste1981) [inline talk/owt/sdk/p2p/p2pclient.cc:274]: You should check if rtc::GetVlueFromJsonObject() returns true before using the signal object. Same for sdp_type parsing.

File: talk/owt/sdk/p2p/p2pclient.cc [hunk 3]
```diff
@@ -251,79 +256,88 @@ void P2PClient::OnSignalingMessage(const std::string& message,
...
           return;
         }
+      }
+    } else if (message_type == "chat-closed") {
+      RTC_LOG(LS_ERROR) << "Handle the situation that PCC is created by "
+                           "received Chat-closed.";
+      int code = 0;
+      std::string error = "";
+      Json::Value stop_info;
+      rtc::GetValueFromJsonObject(json_message, "data", &stop_info);
```
Comment (taste1981) [inline talk/owt/sdk/p2p/p2pclient.cc:307]: ditto.

File: talk/owt/sdk/p2p/p2pclient.cc [hunk 4]
```diff
@@ -251,79 +256,88 @@ void P2PClient::OnSignalingMessage(const std::string& message,
...
+      rtc::GetIntFromJsonObject(stop_info, "code", &code);
+      rtc::GetStringFromJsonObject(stop_info, "message", &error);
+
+      if (code == kWebrtcIceGatheringPolicyUnsupported) {
```
Comment (taste1981) [inline talk/owt/sdk/p2p/p2pclient.cc:311]: Wondering if this is still needed?

File: talk/owt/sdk/p2p/p2pclient.cc [hunk 5]
```diff
@@ -251,79 +256,88 @@ void P2PClient::OnSignalingMessage(const std::string& message,
...
           }
-
           auto new_pcc = that->GetPeerConnectionChannel(remote_id);
+          new_pcc->OnIncomingSignalingMessage(json_message);
           new_pcc->Publish(stream, success_callback, failure_callback);
           return;
         }
+      }
+    } else if (message_type == "chat-closed") {
+      RTC_LOG(LS_ERROR) << "Handle the situation that PCC is created by "
```
Comment (taste1981) [inline talk/owt/sdk/p2p/p2pclient.cc:302]: LS_INFO

File: talk/owt/sdk/p2p/p2pclient.cc [hunk 6]
```diff
@@ -251,79 +256,88 @@ void P2PClient::OnSignalingMessage(const std::string& message,
...
+    Json::Value json_message;
+    if (!reader.parse(message, json_message)) {
+      RTC_LOG(LS_WARNING) << "Cannot parse incoming message.";
+      return;
+    }
+    std::string message_type;
+    rtc::GetStringFromJsonObject(json_message, kMessageTypeKey, &message_type);
     if (!that->IsPeerConnectionChannelCreated(remote_id)) {
-      if (message.find("\"type\":\"chat-closed\"") != std::string::npos) {
+      if (message_type == "chat-closed") {
```
Comment (taste1981) [inline talk/owt/sdk/p2p/p2pclient.cc:268]: please move all the protocol related literals, like "chat-closed", "offer" to the const string listing part of this file.

File: talk/owt/sdk/p2p/p2pclient.cc [hunk 7]
```diff
@@ -251,79 +256,88 @@ void P2PClient::OnSignalingMessage(const std::string& message,
...
       return;
     }
+    Json::Reader reader;
+    Json::Value json_message;
+    if (!reader.parse(message, json_message)) {
+      RTC_LOG(LS_WARNING) << "Cannot parse incoming message.";
+      return;
+    }
+    std::string message_type;
+    rtc::GetStringFromJsonObject(json_message, kMessageTypeKey, &message_type);
```
Comment (taste1981) [inline talk/owt/sdk/p2p/p2pclient.cc:266]: validate the return result.

File: talk/owt/sdk/p2p/p2pclient.cc (hunk 1 above)
Comment (taste1981) [reply talk/owt/sdk/p2p/p2pclient.cc:289]: Adding @dvrogozh

**Comment by taste1981** [comment]
> P2PClient scans signaling messages to see if a PeerConnectionChannel is needed to be created or not. However, scaning messages with std::string::find may miss some matches when extra spaces are added, although these spaces don't change the meaning of a JSON message, e.g.: `{"type":"chat-closed"}` is the same as `{ "type" : "chat-closed" }`. Parsing signaling messages in P2PClient identifies message type more accurate.

Adding @dvrogozh

File: talk/owt/sdk/p2p/p2pclient.cc (hunk 4 above)
Comment (jianjunz) [reply talk/owt/sdk/p2p/p2pclient.cc]: It looks like a workaround for mismatches of gathering policy. Added by https://github.com/open-webrtc-toolkit/owt-client-native/commit/09a71ae9df4a784007a64390e0570d278b33a8b3. Searched kWebrtcIceGatheringPolicyUnsupported and error code 2601, no one else is using it.

File: talk/owt/sdk/p2p/p2pclient.cc (hunk 6 above)
Comment (jianjunz) [reply talk/owt/sdk/p2p/p2pclient.cc]: Done.

**Comment by jianjunz** [comment]
Please also review #719, which makes onPeerConnectionClosed align with what implied by its name. Thanks.

File: talk/owt/sdk/p2p/p2pclient.cc (hunk 2 above)
Comment (jianjunz) [reply talk/owt/sdk/p2p/p2pclient.cc:273]: Same as line 266. It briefly scans the message here to handle some special cases. We only care about messages have "data" filed. Other messages will be dispatched to pcc. pcc is the place for processing signaling messages.

File: talk/owt/sdk/p2p/p2pclient.cc (hunk 7 above)
Comment (taste1981) [reply talk/owt/sdk/p2p/p2pclient.cc:265]: Ack.

**Review by taste1981 — APPROVED** [review]
LGTM.

**Comment by dvrogozh** [comment]
@jianjunz : appreciate if you cab rebase this PR on top of latest master which has merged #719, so I will be able to try them both at once.

**Pull request merged**